test/IdRegistry/IdRegistry.invariants.t.sol

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.18;
+
+import {IdRegistryTestSuite} from "./IdRegistryTestSuite.sol";
+import {IdRegistryHandler} from "./handlers/IdRegistryHandler.sol";
+
+/* solhint-disable state-visibility */
+
+contract IdRegistryInvariants is IdRegistryTestSuite {
+    IdRegistryHandler handler;
+
+    function setUp() public override {
+        super.setUp();
+        idRegistry.disableTrustedOnly();
+        handler = new IdRegistryHandler(idRegistry, address(this));
+
+        targetContract(address(handler));
+    }
+
+    /// @notice All recovery addresses must be associated with a fid.
+    function invariant_1_allRecoveryAddrsAssociatedWithFid() public {
+        address[] memory recoveryAddrs = handler.recoveryAddrs();
```

Comment:
The tricky part about using the handler here is that it is duplicating the logic of the contract to allow testability, but this might not protect against someone introducing a bug in the contract logic later, or if there are bugs in reimplementing the contract logic in the handler. 

Consider the case where [this line](https://github.com/farcasterxyz/contracts/blob/78c205a7e0ab26de151e43c7ccf0198f968ddd29/src/IdRegistry.sol#L279) was accidentally removed, the invariant tests would still pass.

Reply:
Yeah, I agree. The amount of accounting we have to do in the handler is not ideal and a bit of a code smell. I'm going to try replacing some of the state in the handler with stateless calculations. For example, since we know the latest fid, we can iterate over all fids and build up `_fidsByRecoveryAddr` without needing to duplicate the accounting. I think we can do the same for the reverse fid->owner mapping: rather than keep track of it in the handler, we can iterate over all "known addresses" and construct it on demand.